scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontend.java

```diff
@@ -433,21 +410,100 @@ public JobId submit(Job userJob)
+            final InternalJob job = frontendState.createJob(userJob, ident);
 
-            schedulingService.submitJob(job);
-
-            frontendState.jobSubmitted(job, ident);
-            return job.getId();
+            final Future<?> submitting = submitJob(userJob, ident, job);
+            return blockUntilJobIdInitialized(userJob, job, submitting);
         } catch (Exception e) {
             logger.warn("Error when submitting job.", e);
             throw e;
         }
     }
 
+    private Future<?> submitJob(Job userJob, UserIdentificationImpl ident, InternalJob job) {
+        return submitPool.submit(() -> {
+            try {
+                schedulingService.submitJob(job);
+                frontendState.jobSubmitted(job, ident);
+            } catch (Exception e) {
+                String message;
+                if (jobIdInitialized(job)) {
+                    message = String.format("Error while submitting job name: %s; id: %s.",
+                                            userJob.getName(),
+                                            job.getId().value());
+                } else {
+                    message = String.format("Error while submitting job name: %s. JobID wasn't available yet.",
+                                            userJob.getName());
+                }
+                logger.error(message, e);
+                throw new RuntimeException(message, e);
+            } finally {
+                unregisterSumbittedJob(job.getId());
+            }
+        });
+    }
+
+    private boolean finishedWithoutException(final Future<?> submitting)
+            throws InterruptedException, JobCreationException {
+        try {
+            submitting.get();
+            return true;
+        } catch (ExecutionException execException) {
+            // if future is done but it got an exception -> throw exception
+            throw new JobCreationException(execException);
+        }
+    }
+
+    private boolean jobIdInitialized(final InternalJob job) {
+        return job.getId().longValue() != InternalJob.UNINITIALIZED_JOB_ID;
+    }
+
+    /**
+     * This method blocks until we got job id, or exception occurred.
+     * <p>
+     * if future is done but it got an exception -> throw exception
+     * if future is done but no exception occurred
+     * if id is not initialized throw exception, it means we have some severe bug
+     * if id is initialized it is fine we return id
+     * if future is not done
+     * if id is initialized -> register as waiting and return id
+     * if id is not initialize we continue sleep
+     */
+    private JobId blockUntilJobIdInitialized(Job userJob, InternalJob job, Future<?> submitting)
+            throws JobCreationException {
+        try {
+            while (true) {
+                if (submitting.isDone()) {
+                    if (finishedWithoutException(submitting)) {
+                        if (jobIdInitialized(job)) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this expression which always evaluates to "true" [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2589)